typing/typemod.ml

```diff
@@ -85,33 +87,112 @@ let extract_sig_open env loc mty =
 
 (* Compute the environment after opening a module *)
 
-let type_open_ ?used_slot ?toplevel ovf env loc lid =
-  let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
-  match Env.open_signature ~loc ?used_slot ?toplevel ovf path env with
-  | Some env -> path, env
-  | None ->
-      let md = Env.find_module path env in
-      ignore (extract_sig_open env lid.loc md.md_type);
-      assert false
+let type_module_fwd : (Env.t -> Parsetree.module_expr ->
+                       Typedtree.module_expr) ref =
+  ref (fun _ _ -> assert false)
+
+let mod_ident_counter = ref 0
+let generated_module_ident = ref []
+
+let push_current_mid (mi, md, env) =
```

Comment:
As said elsewhere, I'd suggest to try removing this stack and use the code structure to pass the information around.

In addition, the global references should be reset when starting type-checking of a new unit / toplevel phrase.  For mod_ident_counter/generated_module_ident, it might be ok not to do it, but for open_struct_level, this seems required (for a toplevel phrase).

Reply:
Alternatively, turn open_struct_level into just a Boolean reference, and update it with Misc.protect_refs.

Reply:
Good catch. I am using `Misc.protect_refs` now.

See [this comment](https://github.com/ocaml/ocaml/pull/1506#discussion_r162798694) for why we can't turn `open_struct_level` into just a boolean value.